lib/webpage_cache_service/cache.rb

```diff
@@ -0,0 +1,120 @@
+require "tempfile"
+
+class WebpageCacheService
+  class Cache
+    extend Forwardable
+
+    attr_reader :offline_cache
+    attr_accessor :headers
+
+    def initialize webpage:
+      @webpage = webpage
+      @offline_cache = webpage.offline_caches.create
+    end
+
+    def_delegator :@webpage, :uri
+
+    def errors
+      @errors ||= Errors.new
+    end
+
+    def headers
+      @headers ||= DEFAULT_HEADERS.dup
+    end
+
+    def set_header key, value
+      headers[ key ] = value
+    end
+
+    def clear_header key
+      headers.delete key
+    end
+
+    def exec
+      cache
+
+      # Return self since there is an errors object and the offline cache model
+      # that people might care about on this object.
+      self
+    end
+
+    private
+
+    def cache
+      response, root_attachment = get uri: uri
+      @root_attachment = root_attachment
+
+      offline_cache.tap do |obj|
+        obj.root = @root_attachment
+        obj.save
+      end
+
+      if response&.status > 399
+        errors.add uri, "Got non-okay status back from the server: #{ response&.status }"
+        return
+      end
+
+      return unless PARSABLE_MIMES.include? root_attachment.blob.content_type
+
+      links.each { |link| get uri: link }
+    end
+
+    # this has some serious memory implications since it reads the whole file
+    # in, but hopefully people don't have gigabyte sized HTML pages
+    def nokogiri
+      @nokogiri ||= Nokogiri::HTML(@root_attachment.blob.download)
+    end
+
+    def links
+      @links ||= ASSET_XPATHS.flat_map(&nokogiri.method(:xpath))
+        .map(&:to_s)
+        .map(&Addressable::URI.method(:parse))
+        .map { |link| uri + link }
+    end
+
+    def client
+      @client ||= HTTP.headers(headers).follow
+    end
+
+    def get uri:
+      response = client.get uri
+
+      content_type = response.content_type.mime_type
+
+      temp_key = Digest::SHA256.hexdigest uri.to_s
+      attachment = Tempfile.create temp_key do |temp_file|
+        temp_file.binmode
+
+        response.body.each do |partial|
+          temp_file.write partial
+        end
+
+        temp_file.rewind
+
+        unless content_type
+          # some websites *cough*offline.pink*cough* don't return a content type
```

Comment:
<3

Reply:
so it turns out this also got implemented in ActiveStorage the other week https://github.com/rails/rails/commit/c2ba530c43244b5b60fd629f61cd8b44c43ecda9#diff-123aa7e5579a68c3dc535779855c51e1 so whenever that gets released in another RC or something for 5.2 I can rip this out :)

Reply:
<3fancy